### ISSUE TITLE
-Wundefined-reinterpret-cast not triggered if temporary pointer is created?

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#58814]
Hi,

We find that the `-Wundefined-reinterpret-cast` warning triggers only when the dereferencing happens in the same line as the `reinterpret_cast`, but not if an intermediate pointer is created:

```cpp
void foo()
{
    int a;
    float b;
    *reinterpret_cast<float*>(&a) = b;

    auto p = reinterpret_cast<float*>(&a);
    *p = b;
}
```
Repro: https://godbolt.org/z/j3f8KWWPf

I'm curious to understand if that is intended or a bug? Thanks!

**Comment by carlosgalvezp** [comment]
Sounds good, thanks for the clarification and quick response!
